src/main/java/org/jenkinsci/plugins/workflow/actions/QueueItemAction.java

```diff
@@ -0,0 +1,76 @@
+package org.jenkinsci.plugins.workflow.actions;
+
+import hudson.model.InvisibleAction;
+import hudson.model.Queue;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+
+/**
+ * Records information for a {@code node} block.
+ */
+public abstract class QueueItemAction extends InvisibleAction implements PersistentAction {
+    private static final long serialVersionUID = 1;
+
+    /**
+     * Possible queue stats for the item associated with this {@link FlowNode}.
+     */
+    public enum QueueState {
+        QUEUED,
+        CANCELLED,
+        LAUNCHED,
+        UNKNOWN
+    }
+    
+    /**
+     * Gets the {@link Queue.Item} for this task, if it exists.
+     *
+     * @return The item, or null if it's not in the queue.
+     */
+    @CheckForNull
+    public abstract Queue.Item itemInQueue();
+
+    /**
+     * Get the current {@link QueueState} for a {@link FlowNode}. Will return {@link QueueState#UNKNOWN} for
+     * any node without one of an {@link QueueItemAction} or {@link WorkspaceAction}.
+     *
+     * @param node A non-null {@link FlowNode}
+     * @return The current queue state of the flownode.
+     */
+    @Nonnull
+    public static QueueState getNodeState(@Nonnull FlowNode node) {
+        WorkspaceAction workspaceAction = node.getPersistentAction(WorkspaceAction.class);
+        if (workspaceAction != null) {
+            return QueueState.LAUNCHED;
+        } else {
+            QueueItemAction action = node.getPersistentAction(QueueItemAction.class);
+            if (action != null) {
+                Queue.Item item = action.itemInQueue();
+                if (item == null) {
+                    // The item has left the queue completely and there isn't a WorkspaceAction.
+                    return QueueState.CANCELLED;
```

Comment:
Or it is actually being launched, but just spent >5m in `PlaceholderExecutable.run` prior to attaching `WorkspaceAction`. Possible but seems unlikely.